### ISSUE TITLE
react-native: The shadow-cljs Websocket was disconnected.

### ISSUE THREAD

**Issue by MareinK** [thheller/shadow-cljs#957]
Hi, thanks for your work on shadow-cljs!

I've been trying for a while to get shadow-cljs working with Expo. Unfortunately I keep getting the error `The shadow-cljs Websocket was disconnected.` which I can't seem to solve.

When I use `shadow-cljs` to watch the app and then use `expo` to run it on my Android device, I get this error in the `expo` console:

```
The shadow-cljs Websocket was disconnected.
at node_modules/expo/build/logs/LogSerialization.js:156:14 in _captureConsoleStackTrace
at node_modules/expo/build/logs/LogSerialization.js:41:26 in serializeLogDataAsync
- ... 9 more stack frames from framework internals

shadow-cljs - remote-error, Event {
  "isTrusted": false,
  "message": "Failed to connect to localhost/127.0.0.1:9630",
}
at node_modules/expo/build/logs/LogSerialization.js:156:14 in _captureConsoleStackTrace
at node_modules/expo/build/logs/LogSerialization.js:41:26 in serializeLogDataAsync
- ... 9 more stack frames from framework internals
```

The `shadow-cljs - remote-error` section is also shown as a pop-up on the device itself.

To set-up the project, I have mainly been cloning the supposedly-good https://github.com/PEZ/rn-rf-shadow and then trying a number of things from there, including the below, but the error remains.
- use different versions of `shadow-cljs`, including `2.15.3`, `2.15.5`, `2.16.0`, `2.16.2`
- use the `local-ip` setting mentioned on [the REPL troubleshoot page](https://github.com/shadow-cljs/shadow-cljs.github.io/blob/master/docs/repl-troubleshoot.adoc)
- bump `expo-cli` version to 4.13.0
- `expo upgrade` to version 43

I've also set up a project from scratch with the same result.
I'm developing on Linux.

Am I getting something wrong here, or is shadow-cljs misbehaving? I'd really like to develop for expo using cljs and shadow, so any tips are appreciated. Thanks! :)

**Comment by thheller** [comment]
Please verify that you are actually starting the shadow-cljs version you think you are starting. `2.16.+` should defintely have the fixed version and not attempt to connect to `localhost`. You can see the version that is used when starting shadow-cljs. It prints something like

```
shadow-cljs - server version: 2.6.13 running at http://localhost:9630
```

For example just changing the version in `package.json` does not actually change the version, you'd need to run `npm install` after doing so. Or if you are using `deps.edn` or `project.clj` you need to update the version in those instead of `package.json`.

**Comment by MareinK** [comment]
Thanks! Yes, I did run `npm install` already. Here's the output of `npx shadow-cljs watch app` using version 2.16.2:

```
shadow-cljs - config: /home/marein/temp/rn-rf-shadow/shadow-cljs.edn
shadow-cljs - server version: 2.16.2 running at http://localhost:9630
shadow-cljs - nREPL server started on port 38939
shadow-cljs - watching build :app
[:app] Configuring build.
[:app] Compiling ...
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[:app] Build completed. (147 files, 0 compiled, 0 warnings, 11.32s)
```

Thus far I didn't investigate the SLF4J error very much, thinking that would not be the cause of the shadow-cljs problem, but correct me if I'm wrong.

**Comment by thheller** [comment]
The SLF4J error likely comes from one of your `:dependencies` but it shouldn't affect anything in the build so safe to ignore.

And your build config is? I updated it so that `:local-ip` should never be necessary so it should just work without.

Can you run `shadow-cljs clj-repl` and then `(shadow/get-server-addrs)`? You should get a list of all known local ips on your machine (eg. `("192.168.178.5" "172.17.144.1" "172.31.112.1")`). The build will attempt to connect to each until one works. Please verify that the IP your app could reach is in there.

**Comment by MareinK** [comment]
My `shadow-cljs.edn` file looks like this (the default from https://github.com/PEZ/rn-rf-shadow):

```
{:source-paths ["src/main"
                "src/test"]

 :dependencies [[reagent "1.1.0"]
                [re-frame "1.2.0"]]

 :builds       {:app {:target           :react-native
                      :init-fn          example.app/init
                      :output-dir       "app"
                      :compiler-options {:infer-externs :auto}
                      :devtools         {:autoload true
                                         :preloads [shadow.expo.keep-awake]}}}}
```

The result of `(shadow/get-server-addrs)` is:

```
("10.10.2.77")
```

That is what I was expecting, as it is my known local IP.

Maybe this output from `expo` is also useful. It seems to use the above IP in some places, but `localhost` in at least one place:

```
Starting project at /home/marein/temp/rn-rf-shadow
Developer tools running on http://localhost:19002
Starting Metro Bundler

[... QR Code ...]

 › Waiting on exp://10.10.2.77:19000
 
[... command key listing ...]

Logs for your project will appear below. Press Ctrl+C to exit.

[... press 'a' ...]

› Opening on Android...
› Opening exp://10.10.2.77:19000 on moto_g_8__plus

```

Thanks again!

**Comment by thheller** [comment]
The expo servers don't have anything to do with the websocket shadow-cljs attempts to connect.

Your config and options look fine to me. Can you open `http://10.10.2.77:9630` in a webbrowser on the device?

**Comment by MareinK** [comment]
Yes, that seems to correctly show the dashboard.

**Comment by thheller** [comment]
Sorry, I'm out of ideas. Everything looks fine. If you share the repo I can maybe take a look locally and see if something doesn't fit. It has been a couple years since I last tried expo so maybe something in their tooling changed that now causes problems?

**Comment by MareinK** [comment]
It would be great if you want to have a look! For all testing in this issue, I've been using a clone of https://github.com/PEZ/rn-rf-shadow with `"shadow-cljs": "^2.16.2"`. Thanks!

**Comment by thheller** [comment]
That repo works just fine for me.

**Comment by MareinK** [comment]
That's what I feared, as I wouldn't expect the repo to be faulty... Thanks for testing. I guess then that the problem has to be with something else in my environment, like my computer hardware, my OS, or my phone?

I'll probably be trying a different route using figwheel, which is also documented in [the Expo docs](https://docs.expo.dev/guides/using-clojurescript/).

**Comment by thheller** [comment]
I don't have any guesses anymore. The IP detection seems to be fine and that it what it should be using to connect. I don't know how you end up connecting to localhost. That was a bug that was fixed and others have confirmed it fixed as well.

Did you try using `:local-ip "10.10.2.77"` in the build config? It shouldn't be necessary but it might still help?

**Comment by ag91** [comment]
uhm, this issue is similar to what I am experiencing. The only difference seems to be that mine happens only if you start the React Native debugger in Android (iOS works fine).
Steps to reproduce:
1. start the app with react-native run android
2. run in terminal `adb shell input keyevent 82`
3. select Debugger

At this point when Metro reloads the bundle, the connection with shadow-cljs is basically dropped: the REPL cannot reach it anymore. Strangely the existing console.log reach the Chrome debugger window.
As soon as you deactivate the debugger (by repeating 2 and 3), the connection is recovered.

If @MareinK does not have the debugger active, I will open another issue.

**Comment by MareinK** [comment]
> Did you try using `:local-ip "10.10.2.77"` in the build config? It shouldn't be necessary but it might still help?

I did try it, it didn't make a difference.

---

> If @MareinK does not have the debugger active, I will open another issue.

I've tested with and without the debugger open (on `localhost`). For me the error also appeared when the debugger was not active.

---

However, I bring good tidings! A while ago I created [an issue for this problem at the `create-expo-cljs-app` repo](https://github.com/jgoodhcg/create-expo-cljs-app/issues/2#issuecomment-968160637), before I had concluded that the issue was with `shadow-cljs`. A few days ago @chrisghost commented the following on that issue:

> I fixed this by opening the debugger tab using my LAN IP : http://192.168.0.1:19000/debugger-ui/

This workaround seems to work consistently for me! As long as the debugger is open at the local IP (so in my case `10.10.2.77`), `shadow-cljs` works fine and the error does not appear.

I'm very happy as I can finally continue working on my app using `shadow-cljs` and `expo`.

Perhaps this workaround also holds some information as to the cause of the problem.

**Comment by ag91** [comment]
@thheller I confirm the above: using the IP address used by shadow-cljs (the one you see with the --verbose option) instead of localhost makes my Debugger connect and work on Android: maybe there is a wrong IP mapping for localhost?

thanks for sharing the workaround @MareinK !
